tests/cpython/__init__.py

```diff
@@ -0,0 +1,77 @@
+import os
+import json
+import doctest
+import unittest
+
+from test import support
+
+# import json with and without accelerations
+cjson = support.import_fresh_module('json', fresh=['_json'])
+pyjson = support.import_fresh_module('json', blocked=['_json'])
+hyperjson = support.import_fresh_module('hyperjson')
+# JSONDecodeError is cached inside the _json module
+cjson.JSONDecodeError = cjson.decoder.JSONDecodeError = json.JSONDecodeError
+
+# create two base classes that will be used by the other tests
+class PyTest(unittest.TestCase):
+    json = pyjson
+    loads = staticmethod(pyjson.loads)
+    dumps = staticmethod(pyjson.dumps)
+    JSONDecodeError = staticmethod(pyjson.JSONDecodeError)
+
+@unittest.skipUnless(cjson, 'requires _json')
+class CTest(unittest.TestCase):
+    if cjson is not None:
+        json = cjson
+        loads = staticmethod(cjson.loads)
+        dumps = staticmethod(cjson.dumps)
+        JSONDecodeError = staticmethod(cjson.JSONDecodeError)
+
+@unittest.skipUnless(hyperjson, 'requires hyperjson')
+class RustTest(unittest.TestCase):
+    if hyperjson is not None:
+        json = hyperjson
+        loads = staticmethod(hyperjson.loads)
+        dumps = staticmethod(hyperjson.dumps)
+        # FIXME: hyperjson does not have this
+        JSONDecodeError = staticmethod(pyjson.JSONDecodeError)
+        # JSONDecodeError = staticmethod(hyperjson.JSONDecodeError)
+
+# test PyTest and CTest checking if the functions come from the right module
+class TestPyTest(PyTest):
+    def test_pyjson(self):
+        self.assertEqual(self.json.scanner.make_scanner.__module__,
+                         'json.scanner')
+        self.assertEqual(self.json.decoder.scanstring.__module__,
+                         'json.decoder')
+        self.assertEqual(self.json.encoder.encode_basestring_ascii.__module__,
+                         'json.encoder')
+
+class TestCTest(CTest):
+    def test_cjson(self):
+        self.assertEqual(self.json.scanner.make_scanner.__module__, '_json')
+        self.assertEqual(self.json.decoder.scanstring.__module__, '_json')
+        self.assertEqual(self.json.encoder.c_make_encoder.__module__, '_json')
+        self.assertEqual(self.json.encoder.encode_basestring_ascii.__module__,
+                         '_json')
+
+class TestRustTest(RustTest):
+    def test_hyperjson(self):
+        # FIXME: hyperjson does not have any of this
+        pass
+        # self.assertEqual(self.json.scanner.make_scanner.__module__, 'hyperjson')
```

Comment:
No clue what these scanner things are. I guess we should document that incompatibility somewhere. 🤔